### PR TITLE
Implement Wildcard support

### PR DESCRIPTION
**Pull request by JavaScriptDude**
1) Significant rewrite of logic to enable more calling usecases. Was able to eliminate "update" code.
2) Removed format indexes as they are redundant in Python3
3) May be a few more tweaks as I went along

### CODE REVIEW

**Comment by JavaScriptDude** [comment]
Thanks for the eyes :)

File: godaddy.py [hunk 1]
```diff
@@ -28,18 +31,22 @@ def _get_zone(domain):
...
-def _get_subdomain_for(domain, zone):
+def _get_subdomain(domain, zone):
     subdomain = domain[0:(-len(zone)-1)]
     return subdomain
 
 
-def _update_dns(domain, token):
+def _get_txt_records(domain):
+    txt_records = client.get_records(zone, record_type="TXT", name=domain)
+    return txt_records
```
Comment (josteink) [inline godaddy.py]: Where do you get "zone" from here? It's not a parameter.

File: godaddy.py [hunk 2]
```diff
@@ -7,15 +7,18 @@
 import time
 import godaddypy
 
+# See readme.md for explanation of this flag
+LE_WILDCARD_SUPPORT = True
+
```
Comment (josteink) [inline godaddy.py]: This not being an environment parameter is honestly quite ridiculous. Please fix 😄

File: godaddy.py [hunk 3]
```diff
@@ -48,21 +55,53 @@ def _update_dns(domain, token):
...
+    verb='add'
+    gd_api = client.add_record
+    
+    if do_update or not LE_WILDCARD_SUPPORT:
+        if do_update or len(_get_txt_records(subdomain)) > 0:
+            if not do_update and domain.find('*') > -1:
+                logger.warn("+ Warning - Wildcard URL dectected, but LE_WILDCARD_SUPPORT is set to False. If valiation fails, please set LE_WILDCARD_SUPPORT to True and run again.")
+            verb='update'
+            gd_api = client.update_record
+
```
Comment (josteink) [inline godaddy.py]: I find this hard to follow.

1. If `do_update` or some stuff
2. THEN If `do_update` or some other stuff
3. THEN If NOT `do_update`...?

I think detecting this error-condition (error configuration from the user really) and reporting on it, is a *nice touch* and definitely something we should keep...

But is there a better way to express this rather than a messy 3-ifs down thingie? Maybe extract to its own function where we can make things a bit more clean and declarative? Maybe like `_should_update(do_update, domain)`...

Also.. I think reporting this error condition is something we should ALWAYS do. 

Then it becomes simpler, like this:

````python
if domain.find("*") > -1 and LE_WILDCARD_SUPPORT == False:
    logger.warn("+ Warning - Wildcard URL dectected, but LE_WILDCARD_SUPPORT is set to False. If valiation fails, please set LE_WILDCARD_SUPPORT to True and run again.")

if _should_update(do_update, domain):
    verb='update'
    gd_api = client.update_record
````

That's *much* cleaner, right?

Comment (JavaScriptDude) [reply godaddy.py]: Yes, pushing the if blocks to a separate function will improve readability without a big increase in size. I will get that done next.

File: godaddy.py (hunk 2 above)
Comment (JavaScriptDude) [reply godaddy.py]: I'm cool with it being in the environment variables for now but long term, would prefer it lie in the dehydrated config which makes it more user friendly. I'll add this tweak.

Comment (josteink) [reply godaddy.py]: Fair enough. I'm happy with that.

I just don't want us to require the user to actively change the module's code in order to use it, because that makes staying up to date with `git pull` and friends much harder than it needs to be.

**Comment by JavaScriptDude** [comment]
Thanks again for the review. I agree with the idea of batching smaller sets of changes into different PR's.

File: godaddy.py (hunk 2 above)
Comment (JavaScriptDude) [reply godaddy.py]: I totally agree.

**Comment by JavaScriptDude** [comment]
I have a new cleaned up version ready to test but I before I submit, I just came up with a design idea to simplify the code more.

If we design the script to allow running under `HOOK_CHAIN=yes` in Dehydrated config. We will have the full list of domains to validate and automatically use the GoDaddy "add" API only when needed. This will cause much less spamming of TXT records and and we can do away with the LE_WILDCARD_SUPPORT flag we are playing with.

However, I am not sure of any cases where users must use `HOOK_CHAIN=no`.

Thoughts?

**Comment by JavaScriptDude** [comment]
Why don't we take this approach:
1) Recommend in readme.md that HOOK_CHAIN must be enabled to support wildcard domains when the user is registering both the domain (foo.com) and the wildcard (*.foo.com)

2) Auto detect HOOK_CHAIN in the code and if it's on, then be smart enough to use add api when it's appropriate 

3) if HOOK_CHAIN is off and wildcard is detected, warn the user that the validation may fail and give directions to remedy. 

This way, we don't need any configurable and dont have to force users to use HOOK_CHAIN.

**Comment by josteink** [comment]
Sounds good to me!

**Comment by JavaScriptDude** [comment]
Code is updated and tested for HOOK_CHAIN on and off. 

In order to detect HOOK_CHAIN mode, I used the number of arguments passed in the call to the deploy_challenge (create_txt_record) hook. If records is greater than 3 then its assumed to be in HOOK_CHAIN.

Another note, I originally had logic to try to detect a (*) in the domain, but that was invalid as the (*) is never passed to the hook calls. I had to remove this moot code.

File: godaddy.py [hunk 4]
```diff
@@ -33,36 +36,78 @@ def _get_subdomain_for(domain, zone):
...
         'type': 'TXT'
     }
 
-    existing_records = client.get_records(zone, record_type="TXT", name=subdomain)
-    if (len(existing_records) == 0):
-        result = client.add_record(zone, record)
-    else:
-        result = client.update_record(zone, record)
+    def __should_add():
+        if do_update: return False
```
Comment (josteink) [inline godaddy.py]: This function signature looks like it's missing quite a few parameters.

File: godaddy.py [hunk 5]
```diff
@@ -33,36 +36,78 @@ def _get_subdomain_for(domain, zone):
...
+    global HOOK_CHAIN, domain_hist
+
     challengedomain = "_acme-challenge." + domain
     logger.info(" + Updating TXT record for {0} to '{1}'.".format(challengedomain, token))
     zone = _get_zone(challengedomain)
     # logger.info("Zone to update: {0}".format(zone))
     subdomain = _get_subdomain_for(challengedomain, zone)
     # logger.info("Subdomain name: {0}".format(subdomain))
-
+
```
Comment (josteink) [inline godaddy.py]: Some excess whitespace here.

File: godaddy.py [hunk 6]
```diff
@@ -33,36 +36,78 @@ def _get_subdomain_for(domain, zone):
...
-        result = client.add_record(zone, record)
-    else:
-        result = client.update_record(zone, record)
+    def __should_add():
+        if do_update: return False
+        if HOOK_CHAIN:
+            if domain in domain_hist:
+                domain_hist.append(domain)
+                return True
+            domain_hist.append(domain)
```
Comment (josteink) [inline godaddy.py]: This can be moved up to `if HOOK_CHAIN:` to avoid duplication.

File: godaddy.py [hunk 7]
```diff
@@ -33,36 +36,78 @@ def _get_subdomain_for(domain, zone):
...
+            return False
+        return True if len(client.get_records(zone, record_type="TXT", name=subdomain)) == 0 else False
```
Comment (josteink) [inline godaddy.py]: Shouldn't we add to `domain_hist` here?

File: godaddy.py [hunk 8]
```diff
@@ -22,6 +22,9 @@
 logger.addHandler(logging.StreamHandler())
 logger.setLevel(logging.INFO)
 
+domain_hist = None
+HOOK_CHAIN = None
```
Comment (josteink) [inline godaddy.py]: Wouldn't it be better to initialize this to an empty list? That way we won't have to worry about that later?

File: godaddy.py [hunk 9]
```diff
@@ -33,36 +36,78 @@ def _get_subdomain_for(domain, zone):
...
 
 
 def create_txt_record(args):
+    global HOOK_CHAIN, domain_hist
+    HOOK_CHAIN = True if len(args) > 3 else False
+    logger.info("HOOK_CHAIN = {}".format(HOOK_CHAIN))
+    if HOOK_CHAIN: 
+        domain_hist=[]
+    else:
+        logger.warn(" + WARNING: le_godaddy_dns is not running in HOOK_CHAIN mode which does not support Wildcard SAN certificates. To get Wildcard support, please set HOOK_CHAIN=\"yes\" in your Dehydrated config file")
```
Comment (josteink) [inline godaddy.py]: Do we *always* wan't to show this warning, or only when there's a problematic configuration detected?

Also: The warning provided to the user is a double-negative. These are hard to understand, so we should definitely try to re-phrase that into a simpler form.

**Comment by josteink** [comment]
This approach looks much better. I guess it's a little annoying that we never get the `(*)`, but it sort of makes sense from a DNS-perspective.

All in all, I like this much better, but as always I've left a few comments on things I think can be improved.

File: godaddy.py (hunk 7 above)
Comment (JavaScriptDude) [reply godaddy.py]: domain_hist will only have one entry if HOOK_CHAIN = False and thus would be redundant if I moved it elsewhere.

File: godaddy.py (hunk 8 above)
Comment (JavaScriptDude) [reply godaddy.py]: As it's only used if HOOK_CHAIN =True, I want to initialize it when we know it will be used. Small but measurable performance improvement when HOOK_CHAIN = False.

File: godaddy.py (hunk 9 above)
Comment (JavaScriptDude) [reply godaddy.py]: I could not find any alternative to notify the user just once. I tried setting ENV var to store a flag that the user was warned, but that did not work.

File: godaddy.py (hunk 6 above)
Comment (JavaScriptDude) [reply godaddy.py]: I'm not sure I understand. If you mean the two append Lines, it can not be done before we test of domains in domain_hist. We could store that test in an intermediary variable,  but we are not winning anything that way. 

Please suggest a new code block and I'll be glad to test it.

File: godaddy.py (hunk 8 above)
Comment (josteink) [reply godaddy.py]: I honestly don't think considering nano-seconds of execution-time matters the slightly when we down the line actually has a `sleep(10)`. Random network latency in the Godaddy REST is going to affect performance more than this bit, by orders of magnitudes.

I'd rather see us optimize for deterministic code which is easy to reason about. If we initialize the value to the form it *should always have when it is being used*, we don't have to worry about if it has been initialized correctly, or where we should do that initialization. Which means much reduced mental overhead, so we can focus on the actual logic of the module and nothing else.

I'll have to call in the saying about "premature optimization" and "the root of all evil" here 😄

File: godaddy.py (hunk 9 above)
Comment (josteink) [reply godaddy.py]: OK fair enough. But we need to change the message.

How about "Dehydrated is running with HOOK_CHAIN disabled. Consider enabling HOOK_CHAIN for wildcard-support and improved performance."?

I think this is much clearer why one may want to enable it, and if the current setup is problematic to user at hand.

File: godaddy.py (hunk 6 above)
Comment (josteink) [reply godaddy.py]: No. You're right. This was clearly me not paying enough attention to the logic at hand.

Sorry about that.

File: godaddy.py (hunk 7 above)
Comment (josteink) [reply godaddy.py]: You can disregard this comment too. It was meant to be considered with my other comment about adding.

That way (I imagined) we could have a single `.append()` at the top without the repetition, and get less branching and a more consistent exection, but like you said, that would invalidate the logic.

File: godaddy.py (hunk 9 above)
Comment (josteink) [reply godaddy.py]: Also: we really should get rid of the `domain_hist=[]` here. It's really out of place and belongs at initialization.

**Comment by josteink** [comment]
All in all, I think you raise some good points.

I disagree about your points w.r.t. performance though. If you clean that bit up, I think may have (finally!) found a good form for wildcard support which should make everyone happy.

Thanks for hanging out in there, with all my opinionated comments and all 😄

File: godaddy.py (hunk 6 above)
Comment (JavaScriptDude) [reply godaddy.py]: No problem.

File: godaddy.py (hunk 9 above)
Comment (JavaScriptDude) [reply godaddy.py]: As the logic only detects HOOK_CHAIN for SAN Certs, I will change the wording to "Dehydrated may be running with HOOK_CHAIN disabled. Consider enabling HOOK_CHAIN for wildcard-support and improved performance."

**Comment by JavaScriptDude** [comment]
Its tested and should be ready to go.

**Comment by josteink** [comment]
Because there’s so many commits, I recommend squashing this down to one commit when merging. 

Apart from that, no more objections. It’s all yours! 🙂

**Comment by JavaScriptDude** [comment]
Sure, I will Squash commits when I merge. 

However, looks like the file changes in this PR needs approval. Files Changed -> Review Changes -> Approve.

Thanks.
